### ISSUE TITLE
Control and run other scripts via Hangouts

### ISSUE THREAD

**Issue by JojoPiARM** [hangoutsbot/hangoutsbot#893]
## Description
I want to ask you, this plugin hangoutsbot supports to run (open) other examples .sh or example .py scripts, send reply info about outcomes from command via hangouts and close this scripts?

If yes, please help me, which hangoutsbot's commands I can use it?

## Output of `/bot version`
2.8.0 
Raspbian Stretch on RPi3.

**Comment by das7pad** [comment]
Take a look at the `spwan` plugin: https://github.com/hangoutsbot/hangoutsbot/wiki/Spawn-Plugin

**Comment by JojoPiARM** [comment]
Thank you for your help. I tried "spawn" plugin more time on the RPi3 and I have a issues. The last attempt caused me a hangupsbot malfunction, and I had to reinstall the whole system. 
I done "**/bot addplugin spawn**" and RPi3 replied: 
add plugin:spawn 
*loaded: plugins.spawn
*added to config.json
I checked "**/bot getplugins**" and RPi3 replied:
9 loaded plugins (config.json)
* autoreply
* botaliases
* default
* dnd
* mentions
* remind
* spawn
* starter
* subscribe
63 available plugins
* _DEPRECATED_simplytranslate
* _DEPRECATED_syncrooms_autotranslate
* _cam_mail_intercept
* _example
* _example.example_memory
* _example.example_printrenames
* _example.example_watchmembers
* _example.example_webbridge
* _unittest
* _unittest.unittest_geticon
* _unittest.unittest_memory
* _unittest.unittest_reprocessor
* _unittest.unittest_shutup
* _unittest.unittest_statusevents
* api
* botaddnotif
* botalive
* catfact
* chance
* chatlogger
* cleverbot
* convtools
* convtools_invitations
* easteregg
* forecast
* forwarding
* foursquare
* humor_hangoutcalls
* image
* image.image_linker_reddit
* image.image_links
* image.image_memegenerator
* image.image_screenshot
* lookup
* lottery
* metar
* monitoradds
* namelock
* restrictedadd
* showme
* simplewikipedia
* slack
* slackrtm
* spotify
* syncrooms
* syncrooms_config
* telesync
* tldr
* twitter
* urbandict
* webbridge_hubot
* webbridge_telegram
* wolframalpha
* xkcd
I tried adding commands in config.json but then stopped working. I do not know which procedure is right. Would you please write instructions with my orders to enter directly into the Hangouts app so config.json was properly configured. So I do not have to manually issue commands to the config.json file. I know two config.json files are stored in the system, so the first config.json serves as a template, and the second config.json user is configured.

**Comment by OllieTerrance** [comment]
The spawn plugin can't be configured at runtime because it needs to register arbitrary commands -- you need to specify it in the config file, then reload config and the plugin.  There should be `spawn` in your list of plugins, and the `spawn` config section should be added as described in the [plugin wiki page](https://github.com/hangoutsbot/hangoutsbot/wiki/Spawn-Plugin).

If you insist on editing your config via the bot, there are [config commands](https://github.com/hangoutsbot/hangoutsbot/wiki/Default-Commands-Plugin#admin-only-commands) you can use, but you will find it easier just to edit the config file yourself -- the default location is `~/.local/share/hangoutsbot/config.json`, unless you've set a different path using the `--config` argument.

**Comment by JojoPiARM** [comment]
If I do understand right, the spawn plugin is solid and I can config only config.json in the default location. I will try it - stopped runing bot, I add spawn config section from plugin wiki page into config.json, then I run hangoutsbot. Is it correct practices? I don't know how to I reload plugin and config before running Hangoutsbot?

**Comment by OllieTerrance** [comment]
> I can config only config.json in the default location

What?  As above, the default can be changed with the `--config` option when starting hangoutsbot.

> Is it correct practices?

You generally don't want or need to stop the bot -- either use the config commands for quick changes, or update the file manually and reload for significant edits.

> how to I reload plugin

See [plugin commands](https://github.com/hangoutsbot/hangoutsbot/wiki/Plugin-Control-Commands#reload-a-plugin).

```
/bot pluginreload plugins.<plugin>
```

> and config

Also listed in [config commands](https://github.com/hangoutsbot/hangoutsbot/wiki/Default-Commands-Plugin#admin-only-commands).

```
/bot reload
```

**Comment by JojoPiARM** [comment]
I tried it and my config.json is /.local/share/hangoutsbot/config.json and I configured spawn on this config.json, then I reloaded with /bot pluginreload plugins.spawn and I received "plugins.spawn: reloaded". I reloaded by /bot reload and it showed a mistake: reload JSONDecodeError: Expecting", delimiter: line 62 column 3 (char 1195). How to I solve it, please?

**Comment by OllieTerrance** [comment]
> JSONDecodeError

You've made a syntax error in your config file.  Luckily you've already got its position in the file:

> line 62 column 3

Also, you should reload your config _before_ the plugin, seeing as the plugin will read its config at startup.

**Comment by JojoPiARM** [comment]
Thank you for your advice, but I try some commands in order to bot reply "Not configured". It is disclosed on the code line 144 on the "spawn.py". I tried ` /bot spawn home` and bot replied "Unknown Command". Please which command I can use it?

**Comment by OllieTerrance** [comment]
> bot reply "Not configured". It is disclosed on the code line 144 on the "spawn.py"

https://github.com/hangoutsbot/hangoutsbot/blob/e4157732f239a5701dd801811690d98134fc5938/hangupsbot/plugins/spawn.py#L142-L145

Then you haven't specified `command` in your spawn entry, meaning the bot has no idea what it's supposed to do when you call the command.

> `/bot spawn home`

If you register a spawn entry called `home`, then the command is simply `/bot home`.

**Comment by JojoPiARM** [comment]
Sorry, I have a many questions, I am trying to do it so it works. It still doesn't work and the words "Not configured" haven't come yet. I have those code in the `config.json` but it doesn't work.

`1. "spawn": {
2.         "home": "/nonstandard/home/directory",
3.         "commands": {
4.             "fortune": {
5.                 "command": ["/usr/games/fortune"]
6.             },
7.             "stock-info": {
8.                 "command": ["/home/portfolio/stock-info", "--"]
9.                 "home": "/home/portfolio",
10.                 "allow_args": true
11.             }
12.         }
13.     }`
Which line can I enter an address to _temp.py_? 
Address is: _home/pi/Jojo_programs/temp.py_

**Comment by JojoPiARM** [comment]
Sorry, I have a many questions, I am trying to do it so it works. It still doesn't work and the words "Not configured" haven't come yet. I have those code in the config.json but it doesn't work.

1. "spawn": { 
2. "home": "/nonstandard/home/directory", 
3. "commands": { 
4. "fortune": { 
5. "command": ["/usr/games/fortune"] 
6. }, 
7. "stock-info": { 
8. "command": ["/home/portfolio/stock-info", "--"] 
9. "home": "/home/portfolio", 
10. "allow_args": true 11. } 12. } 13. }
Which line can I enter an address to temp.py?
Address is: home/pi/Jojo_programs/temp.py

**Comment by JojoPiARM** [comment]
```
    "spawn": {
        "home": "**home/pi/Jojo_programs**",
        "commands": {
            "fortune": {
                "command": ["**python temp.py**"]
            },
            "stock-info": {
                "command": ["/home/portfolio/stock-info", "--"]
                "home": "/home/portfolio",
                "allow_args": true
            }
        }
    }
```
The bold text should be work? Please explain me, can I change word "fortune" to "temp" ?

**Comment by JojoPiARM** [comment]
The bold text I thought as words **home/pi/Jojo_programs** and **python temp.py** in the code at the previous comment.

**Comment by OllieTerrance** [comment]
> can I change word "fortune" to "temp" ?

Instead of asking, have you actually _tried_ it for yourself?

---

> `"command": ["python temp.py"]`

That will try to call an executable called "python temp.py" (that is, a single filename with a space in it), which most likely doesn't exist.  Akin to doing this in a shell:

```
$ "python temp.py"     
bash: python temp.py: command not found
```

What you meant to do was use `command` as a list of executable and arguments, like in a shell:

```
$ python temp.py
This is temp.py.
```

As such, your `command` would be `["python", "temp.py"]`.  If it can't find them, you may need to be explicit with the paths, i.e. `["/usr/bin/python", "/home/pi/Jojo_programs/temp.py"]`.

**Comment by JojoPiARM** [comment]
Yes, I tried it myself.

I tried it last two week and it works. Thank you very much. This `command `  `["/usr/bin/python", "/home/pi/Jojo_programs/temp.py"]` is correct and right. Those command can to use other shell - e.g. shutdown system, show temperature,....

Maybe I will comment, when I will have some issues about Hangoutsbot.
